### DISCUSSION TITLE
Binding Command on TemplatedControls doesn't work

### DISCUSSION THREAD

**Discussion by lucas-miak** [AvaloniaUI/Avalonia#11039] [Q&A] [answered]
**Describe the bug**
Hello, I am trying to create a reusable element called NavButton with TemplatedControl.
But I can't bind an event when I click, no error in the console nor when I click nothing happens.

**Here is my code**

`HomeView.axaml :`

```AXAML
[...]
<components:NavButton Title="Accueil" Icon="fa-house" Command="{Binding HandleSwitchTab}"/>
[...]
```

`HomeViewModel.cs :`

```C#
[...]
public ICommand HandleSwitchTab { get; }

public HomeViewModel(IScreen screen, Client c) {
            [...]
            HandleSwitchTab = ReactiveCommand.Create(
                () => Console.WriteLine("A reactive command is invoked!")
            );
}
[...]      
```

`NavButton.axaml`

```AXAML
<Styles xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:controls="using:CookinGest.Components"

        xmlns:i="clr-namespace:Projektanker.Icons.Avalonia;assembly=Projektanker.Icons.Avalonia"
>

    <Design.PreviewWith>
        <controls:NavButton />
    </Design.PreviewWith>

    <Style Selector="controls|NavButton /template/ Button">
        <Setter Property="Background" Value="#00000000"></Setter>
    </Style>

    <Style Selector="controls|NavButton">
        <!-- Set Defaults -->
        <Setter Property="Template">
            <ControlTemplate>
                <Button Command="{TemplateBinding Command}"
        CommandParameter="{Binding RelativeSource={RelativeSource TemplatedParent}}"> <!-- BEFORE RelativeSource {RelativeSource Self}}-->
                    <Border Height="75" Width="75" CornerRadius="8">

                        <Grid>
                            <Grid.RowDefinitions>
                                <RowDefinition Height="*" />
                                <RowDefinition Height="Auto" />
                            </Grid.RowDefinitions>
    
                            <Grid.ColumnDefinitions>
                                <ColumnDefinition Width="*" />
                            </Grid.ColumnDefinitions>

                            <i:Icon FontSize="20" Value="{TemplateBinding Icon}" Grid.Row="0" Grid.Column="0"/>

                            <TextBlock Text="{TemplateBinding Title}" Grid.Row="1" Grid.Column="0" HorizontalAlignment="Center" />
                        </Grid>
                    </Border>
                </Button>
            </ControlTemplate>
        </Setter>
    </Style>
</Styles>

```

`NavButton.axaml.cs`

```C#

using System.Windows.Input;
using Avalonia;
using Avalonia.Controls.Primitives;

namespace CookinGest.Components;
public partial class NavButton : TemplatedControl
{
    public static readonly StyledProperty<string> TitleProperty = AvaloniaProperty.Register<LargeLabelControl, string>(
        nameof(Title), "<Tab-Name>");

    public string Title
    {
        get => GetValue(TitleProperty);
        set => SetValue(TitleProperty, value);
    }

    public static readonly StyledProperty<string> IconProperty = AvaloniaProperty.Register<LargeLabelControl, string>(
        nameof(Icon), "fa-house");

    public string Icon
    {
        get => GetValue(IconProperty);
        set => SetValue(IconProperty, "fa-solid " + value);
    }

    // Click

    public static readonly StyledProperty<ICommand> CommandProperty =
       AvaloniaProperty.Register<NavButton, ICommand>(nameof(Command));

    public ICommand Command
    {
        get { return GetValue(CommandProperty); }
        set { SetValue(CommandProperty, value); }
    }

    // Constructor

    public NavButton()
    {
        DataContext = this;
    }
}
```

**Desktop :**

- OS: MacOS Ventura 
- Version 13.1 

**Additional context**
I use ReactiveUI

**I'm starting with avalonia and I'm stuck on this problem for a few days, any help would be precious. 
Thank you for your help!**

**Comment by timunie** [comment]
@Lk2v I don't think this is an actual Avalonia bug, so better to open a Q&A for such questions. I'll move this issue into a Q&A. I hope you don't mind.

**Comment by timunie** [comment, marked as answer]
@Lk2v I think the issue you have is this line: 

```cs
    public NavButton()
    {
        DataContext = this;
    }
```

this will set the DataContext for the button to the button itself, which means any binding will lookup the properties inside the control and no the `ViewModel`. See: https://docs.avaloniaui.net/docs/data-binding/the-datacontext

Please try to remove the constructor and try it again. 

HINT: You can also base on `Button` instead of `TemplatedControl`. That way you will already have `Command` and `CommandParameter` as well as `Content` defined. Only `Icon` needs to be added. 
Happy coding

Tim

**Reply by lucas-miak** [reply]
> @Lk2v I think the issue you have is this line: 
> 
> 
> 
> ```cs
> 
>     public NavButton()
> 
>     {
> 
>         DataContext = this;
> 
>     }
> 
> ```
> 
> 
> 
> this will set the DataContext for the button to the button itself, which means any binding will lookup the properties inside the control and no the `ViewModel`. See: https://docs.avaloniaui.net/docs/data-binding/the-datacontext
> 
> 
> 
> Please try to remove the constructor and try it again. 
> 
> 
> 
> HINT: You can also base on `Button` instead of `TemplatedControl`. That way you will already have `Command` and `CommandParameter` as well as `Content` defined. Only `Icon` needs to be added. 
> 
> Happy coding
> 
> 
> 
> Tim

Thank you so much !
Besides, I'm going to follow your advice and work with a Button rather than a templatedControl, that will simplify things.
